core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/messaginglayer/NetworkOutputGate.scala

```diff
@@ -0,0 +1,104 @@
+package edu.uci.ics.amber.engine.architecture.messaginglayer
+
+import akka.actor.{Actor, ActorRef}
+import edu.uci.ics.amber.engine.architecture.messaginglayer.NetworkOutputGate.{
+  NetworkMessage,
+  QueryActorRef,
+  ReplyActorRef
+}
+import edu.uci.ics.amber.engine.common.ambermessage.neo.InternalMessage
+import edu.uci.ics.amber.engine.common.ambertag.neo.Identifier
+
+import scala.collection.mutable
+
+object NetworkOutputGate {
+
+  /** Identifier <-> ActorRef related messages
+    */
+  final case class QueryActorRef(id: Identifier)
+  final case class ReplyActorRef(id: Identifier, ref: ActorRef)
+
+  /** All outgoing message should be eventually NetworkMessage
+    * @param uuid
+    * @param internalMessage
+    */
+  final case class NetworkMessage(uuid: Long, internalMessage: InternalMessage)
+
+  /** Ack for NetworkMessage
+    * note that it should NEVER be handled by the main thread
+    * @param uuid
+    */
+  final case class NetworkAck(uuid: Long)
+}
+
+/** This trait handles the transformation from identifier to actorRef
+  * and also sends message to other actors. This is the most outer part of
+  * the messaging layer.
+  * It SHOULD be a trait since it is highly coupled with actors' sending/receiving logic.
+  * TODO: this trait should eventually become another actor, which is the sender actor.
+  */
+trait NetworkOutputGate {
+  this: Actor => // it requires the class to be an actor.
+
+  private val idMap = mutable.HashMap[Identifier, ActorRef]()
+  private val messageStash = mutable.HashMap[Identifier, mutable.Queue[InternalMessage]]()
+  private var messageUUID = 0L
```

Comment:
It's not a UID. It's a sequence number. And why is the sequence number here? I thought it was in the messaigng layer